packages/ui/src/server/controllers/vcs.ts

```diff
@@ -0,0 +1,320 @@
+import {AppRouteHandler, Request, Response} from '@gravity-ui/expresskit';
+import {VCSSettings} from '../../shared/ui-settings';
+import {AxiosError} from 'axios';
+import {AppConfig} from '@gravity-ui/nodekit';
+import {GithubRepository, GitlabRepository, VcsConfig} from '../../shared/vcs';
+import {sendError} from '../utils';
+import {VcsFactory} from '../components/vcs';
+
+const COOKIE_OPTIONS = {
+    httpOnly: true,
+    secure: true,
+    sameSite: 'strict' as const,
+    maxAge: 1000 * 60 * 60 * 24 * 365,
+};
+
+const ERROR_MESSAGE = {
+    VCS_CONFIG_NOT_FOUND: 'Vcs config is not found',
+    VCS_ID_REQUIRED: 'Vcs id is required',
+    VCS_NOT_SUPPORTED: 'This vcs id is not supported',
+    TOKEN_REQUIRED: 'Token is required',
+    PATH_REQUIRED: 'Path is required',
+    BRANCH_REQUIRED: 'Branch is required',
+    REPOSITORY_REQUIRED: 'Repository is required',
+    UNAUTHORIZED:
+        'The service cannot authorize you. Check the token. You can add new token in the section Settings -> VCS',
+};
+
+const cookieNameByVcsId = (vcsId: string) => `vcs_${vcsId}`;
+
+const loadVcsConfig = (req: Request): VCSSettings[] => {
+    const {uiSettings} = req.ctx.config as AppConfig;
+    return uiSettings.vcsSettings || [];
+};
+
+const sendApiError = (res: Response, error: AxiosError) => {
+    let status = error.response?.status;
+    let message = error.response?.statusText;
+
+    if (status === 401) {
+        status = 500;
+        message = ERROR_MESSAGE.UNAUTHORIZED;
+    }
+
+    sendError(res, message || error.message, status || 500);
+};
+
+export const getVcsConfig: AppRouteHandler = (req, res) => {
+    const cookies = req.cookies;
+    try {
+        const config = loadVcsConfig(req);
+        if (!config.length) {
+            sendError(res, ERROR_MESSAGE.VCS_CONFIG_NOT_FOUND, 404);
+            return;
+        }
+
+        const response = config.map<VcsConfig>((item) => ({
+            ...item,
+            token: cookieNameByVcsId(item.id) in cookies,
+        }));
+        res.status(200).json(response);
+    } catch (e) {
+        req.ctx.logError('Error getting vcs configuration', e);
+        sendError(res, e, 500);
+    }
+};
+
+export const createToken: AppRouteHandler = (req, res) => {
+    const {vcsId, token} = req.body;
+
+    if (!vcsId) {
+        sendError(res, ERROR_MESSAGE.VCS_ID_REQUIRED, 400);
+        return;
+    }
+
+    const config = loadVcsConfig(req);
+
+    if (!config.length) {
+        sendError(res, ERROR_MESSAGE.VCS_CONFIG_NOT_FOUND, 404);
+        return;
+    }
+
+    const vcs = config.find((item) => item.id === vcsId);
```

Comment:
```ts
const vcs = config?.find((item) => item.id === vcsId);
```
and you don't need to check `config.length`, and then `ERROR_MESSAGE.VCS_CONFIG_NOT_FOUND` is also unnecessary